### ISSUE TITLE
from sklearn.datasets import fetch_openml

### ISSUE THREAD

**Issue by hamzza-K** [ageron/handson-ml#416]
mnist = fetch_openml('mnist_784', version=1, cache=True)
mnist.target = mnist.target.astype(np.int8)

whenever i run this piece of snippet in my terminal, my **OS** gets *stuck* for reasons unknown to me.
I'm using **Ubuntu 18.04**.

**P.S** All this code doing is fetching the **mnist** dataset which is also provided in keras dataset. As an alternative, can one use this dataset for this exercise?

**Comment by ageron** [comment]
Hi @hamzza-K ,
Thanks for your feedback.  I noticed that `fetch_openml()` is very slow, so perhaps your OS is not stuck, you just need to wait for a few minutes. If that does not work, then perhaps you can try clearing Scikit-Learn's cache: just remove the `$HOME/scikit_learn_data/openml/openml.org/` directory and try again.  If that still doesn't work, then perhaps there is a proxy or firewall blocking the download. Also, perhaps it's using a lot of RAM, so you don't run out of it (e.g., not too many programs open, and close some tabs in your navigator).
If nothing works, then indeed you can use any other method for downloading MNIST, including using Keras:

```python
from tensorflow import keras
(X_train, y_train), (X_test, y_test) = keras.datasets.mnist.load_data()
```

However, `X_train` will be a 3D array (shape `[60000, 28, 28]`), whereas `fetch_openml()` returns a 2D array (shape `[70000, 784]`), so you need to reshape `X_train` and `X_test`. Plus, it contains `uint8` (unsigned bytes) from 0 to 255, instead of `float64`, so no need to cast to byte (`astype(np.int8)`).

```python
X_train = X_train.reshape(-1, 784)
X_test = X_test.reshape(-1, 784)
```

Also note that `fetch_openml()` returns the full dataset (70,000 images), whereas `load_data()` returns two sets (train=60,000 images + test=10,000 images).
Hope this helps.

**Comment by hamzza-K** [comment]
Thanks ! that clarified a lot and also resolved the issue i had.
Much appreciated.

**Comment by ageron** [comment]
Hi @hamzza-K ,
I'm glad this fixed your issue. In case someone else has the same problem, could you please indicate what exactly the problem was, and what fixed it?
Thanks!
Aurélien

**Comment by hamzza-K** [comment]
from tensorflow import keras
(X_train, y_train), (X_test, y_test) = keras.datasets.mnist.load_data()

X_train = X_train.reshape(-1, 784)
X_test = X_test.reshape(-1, 784)

Since, I had tensorflow installed on my machine, I used keras to
extract the mnist dataset. However, I was negligent of the concept of
arrays being

returned in 3D whilst the code in the book works on 2D. Reshaping the
train and test matrices solved my problem. Sorry I did not try other
two suggestions

that you gave because this seemed more plausible to me.

Thanks again though !

**Comment by blessing-art** [comment]
can we use conda instead of  TensorFlow to import mnist dataset

**Comment by blessing-art** [comment]
from sklearn.datasets import fetch_openml 
mnist = fetch_openml('mnist_784', version=1, cache=True)

I tried this with my window 10, it was prolonged and to even responded. 
can I get another means of downloading the mnist data using anaconda, even if manually?

**Comment by ageron** [comment]
Hi @blessing-art ,

Sadly, `fetch_openml()` is indeed very slow to load MNIST, it's a bit annoying. This is a known issue, due to the way openml stores the data, and the fact that Scikit-Learn uses a pure Python parser (see https://github.com/scikit-learn/scikit-learn/issues/19774 for more details). It will be solved once openml switches to a better representation, which is work in progress.

If you prefer, you can use `tf.keras.datasets.mnist.load_data()` instead: on my machine, it loads the MNIST dataset in 340ms  instead of 45 seconds! However, it's not exactly represented the same way. Here's a table I did a while back (in https://github.com/ageron/handson-ml/issues/301) showing the difference between various methods (two of which are now deprecated). Note that `fetch_openml()` now returns a Pandas DataFrame by default. If you want a NumPy array instead, you need to pass it the `as_frame=False` argument.


| **_Loading MNIST_** | Sklearn `fetch_mldata()` (deprecated)  | Sklearn `fetch_openml()` | TensorFlow `input_data()` (deprecated) | tf.keras `load_data()` |
| ---- | ------------- | ------------- | ------------- | ------------- |
| **Sorted by label** | Yes | No | No | No |
| **Image dtype** | uint8 (0 to 255) | float64 (0.0 to 255.0) | float32 (0.0 to 1.0) | uint8 (0 to 255) |
| **Image shape** | [784] | [784] | [784] | [28, 28] |
| **Label dtype** | float (0.0 to 9.0) | string ('0' to '9') | uint8 (0 to 9) | uint8 (0 to 9) |
| **Split** | No split, but by convention: Train 0..59999, Test 60000...69999 | No split, but by convention: Train 0..59999, Test 60000...69999 | Validation 0..4999, Train 5000..59999, Test 60000..69999 | Train 0..59999, Test 60000..69999 |

I've never tried to install MNIST using conda (e.g., with https://anaconda.org/conda-forge/mnist), but it would certainly work. The only problem is that you would then need to load and parse the files yourself. I prefer using `tf.keras.datasets.mnist.load_data()`.

Hope this helps.
